### DISCUSSION TITLE
Failover VPN (AWS)

### DISCUSSION THREAD

**Discussion by wofferl** [libreswan/libreswan#1770] [Ideas]
With the removal of dpdaction and the changes in 5.0 there was an unofficial feature removed to have a redundant vpn failover with an own updown script.

The trick here was to configure two identical tunnels where one is set to start and the second one is only added initially to avoid "cannot install eroute"-error. If a tunnel goes down (dpdaction=restart) the updown script switches to the second tunnel. This happens until one tunnel is up again.

Because of the changes the connection is now always up and the updown script won't be called during retransmit timeouts or dpd actions.

Paul mentioned in [[Swan] IPsec Failover Multiple Peer Connections to 1 Private IP](https://lists.libreswan.org/pipermail/swan/2023/004281.html) to use mark and priority with 5.0 to set up multiple tunnels with the same subnets.
This works but it solves not the problem, when one tunnel is down, because the active xfrm policy is always set to hold and blocks all traffic. Therefore you need other tools that checks and control the traffic (setting marks, etc.) from outside pluto.

Unfortunately the policy must have different marks or if_id, only different priorities don't work. To avoid having to worry about controlling the traffic with different mark or ipsec interfaces, you can set one with mark and one without.

For example you have mark=0x10 for the second and no mark for the first tunnel. The packets are always marked with 0x10 to match both policies.

```
src 192.0.3.0/24 dst 192.1.4.0/24
        dir out priority PRIORITY ptype main
        tmpl src 192.1.3.33 dst 192.1.2.23
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir fwd priority PRIORITY ptype main
        tmpl src 192.1.2.23 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir in priority PRIORITY ptype main
        tmpl src 192.1.2.23 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
src 192.0.3.0/24 dst 192.1.4.0/24
        dir out priority PRIORITY ptype main
        mark 0x10/0xffffffff
        tmpl src 192.1.3.33 dst 192.1.2.45
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir fwd priority PRIORITY ptype main
        mark 0x10/0xffffffff
        tmpl src 192.1.2.45 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir in priority PRIORITY ptype main
        mark 0x10/0xffffffff
        tmpl src 192.1.2.45 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
```

But if the tunnel with the same or higher priority goes down the policy blocks the traffic nevertheless.
```
src 192.0.3.0/24 dst 192.1.4.0/24
        dir out priority PRIORITY ptype main
        mark 0x10/0xffffffff
        tmpl src 0.0.0.0 dst 0.0.0.0
                proto esp reqid 0 mode transport
src 192.0.3.0/24 dst 192.1.4.0/24
        dir out priority PRIORITY ptype main
        tmpl src 192.1.3.33 dst 192.1.2.23
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir fwd priority PRIORITY ptype main
        tmpl src 192.1.2.23 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
src 192.1.4.0/24 dst 192.0.3.0/24
        dir in priority PRIORITY ptype main
        tmpl src 192.1.2.23 dst 192.1.3.33
                proto esp reqid REQID mode tunnel
```

So I came up with different ideas how to solve this:

1) Calling updown during reviving/dpd with a new PLUTO_VERB as replacement for the old dpdaction=restart. This would restore the old possibility to have one tunnel up and switch dynamically between them if one goes down.

2) A new option to automatically change priority during renegotiation. Decreasing the priority during the revive when setting both tunnels to the same initially priority would solve the problem. This could be controlled with a new option.

2) Adding a new option like negotiationshunt=failover or =none which removes the policy during reviving, which can then be set on one tunnel, with the disadvantage that if the second tunnel is not actually added you can leak packets.

3) Something like auto=failover and linking two connections together will probably be too complex.

But if [Multiple_tunnels_fail_with_Amazon's_VPN](https://libreswan.org/wiki/Interoperability#Multiple_tunnels_fail_with_Amazon's_VPN) still applies, then only the first one will work anyway.

Maybe I'm missing something and there is already another possibility.

**Comment by wofferl** [comment]
@paulwouters @cagney don't know if this was the right place to discuss this topic and if you read the discussions periodically, but before I change or implement somthing regarding this, I want to make sure it could accepted officially.

**Comment by cagney** [comment]
I've turned it into a bug :-)
